drivers/pulse_counter/pulse_counter.c

```diff
@@ -0,0 +1,67 @@
+/*
+ * Copyright (C) 2017 UC Berkeley
+ *
+ * This file is subject to the terms and conditions of the GNU Lesser
+ * General Public License v2.1. See the file LICENSE in the top level
+ * directory for more details.
+ *
+ */
+
+/**
+ * @ingroup     drivers_pulse_counter
+ * @{
+ *
+ * @file
+ * @brief       Driver for the PULSE COUNTER.
+ *
+ * @author      Hyung-Sin Kim <hs.kim@cs.berkeley.edu>
+ *
+ * @}
+ */
+
+#include <string.h>
+
+#include "pulse_counter_params.h"
+
+#define ENABLE_DEBUG    (0)
+#include "debug.h"
+
+/**
+ * @brief   Define the number of configured sensors
+ */
+#define PULSE_COUNTER_NUM    (sizeof(pulse_counter_params)/sizeof(pulse_counter_params[0]))
+
+uint16_t pulse_count[PULSE_COUNTER_NUM];
+
+/* Accumulate pulse count */
+void pulse_counter_trigger(void* arg) {
+    pulse_counter_t *dev = (pulse_counter_t*)arg;
+    pulse_count[dev->id]++;
+}
+
+/* Initialize pulse counter */
+int pulse_counter_init(pulse_counter_t *dev, const pulse_counter_params_t *params)
+{    
+    dev->p.gpio = params->gpio;
+    pulse_counter_reset(dev);
+    if (gpio_init_int(params->gpio, GPIO_IN_PU, GPIO_FALLING, pulse_counter_trigger, dev)) {
+        return -1;
+    }
+    return 0;
+}
+
+/* Return the accumulated pulse counts and reset the count to zero */
+int16_t pulse_counter_read(const pulse_counter_t *dev)
+{
+    int16_t pulse_count_output = pulse_count[dev->id];
+    pulse_counter_reset(dev);
```

Comment:
This is non-atomic, if a pulse arrives between the line above and this line, the pulse will be lost.

A safer implementation is to use atomic_exchange (see my previous comment for an implementation suggestion).

Reply:
...or you could remove the reset when read functionality as suggested by @cladmi. I only wanted to put the idea out there about a reset when read counter.